### PR TITLE
feat: filters (categories and authors) endpoint

### PR DESCRIPTION
**Pull request by AlfreMK**
### 🛠 Changes being made

- Add authors endpoints for filters in frontend usage.
- Add categories endpoint for filters in frontend usage.
- Add book recommendations by a single book endpoint.
- Remove not found exception in get books given that is better for the frontend to get an empty array than an error.

### 🏎 Quality check

- [x] Are your changes following SOLID principles?

- [x] Did you make sure that there aren't any erroneous console logs, debuggers or leftover code in your changes?

- [x] Walk away, take a break, re-read what you filled out above does it make sense if you were coming in cold? What extra context could you provide?

### CODE REVIEW

File: src/author/author.service.ts [hunk 1]
```diff
@@ -1,16 +1,79 @@
...
+      JOIN books ON authors.id = books."authorId"
+      WHERE books."isDeleted" = false
+      GROUP BY authors.id
+      ORDER BY booksCount DESC
+      LIMIT 10
+    `
+
+    // Given that the count of books is a bigint, we need to do this
+    // custom conversion to avoid JSON serialization issues
+    return json(authors)
```
Comment (BMarink512) [inline src/author/author.service.ts:71]: ok

File: src/author/author.controller.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    description: 'List of authors',
+    type: AuthorsResponse,
+  })
+  async getAllCategories(@Query() query: GetAuthorsDto) {
+    const { page, limit, ...filters } = query
+    const books = await this.authorService.getAllAuthors(page, limit, filters)
+    return books
+  }
+
+  @Get('most-books')
```
Comment (BMarink512) [inline src/author/author.controller.ts:24]: This endpoint has no docs

File: src/book/book.service.ts [hunk 3]
```diff
@@ -68,6 +69,11 @@ export class BookService {
       }
     }
 
+    // Given that we get a string from the query, we need to convert it to a boolean
+    if (isDeleted === 'true' || isDeleted === 'false') {
```
Comment (BMarink512) [inline src/book/book.service.ts]: I suggest that the related logic to the filterconditions move it to a separate function called setFilters

File: src/book/book.service.ts [hunk 4]
```diff
@@ -123,6 +130,38 @@ export class BookService {
     return book
   }
 
+  async getBookRecommendationsByISBN(isbn: string) {
```
Comment (BMarink512) [inline src/book/book.service.ts:107]: good

File: src/category/category.service.ts [hunk 5]
```diff
@@ -37,4 +37,41 @@ export class CategoryService {
...
+    page: number = 1,
+    limit: number = 50,
+    filter: CategoryFilterDto = {},
+  ) {
+    const { name } = filter
+
+    const filterConditions: any = {}
+
+    if (name) {
+      filterConditions.name = { contains: name, mode: 'insensitive' }
```
Comment (BMarink512) [inline src/category/category.service.ts]: Similar comment as the book service, maybe it would be best to have a shared folder or utils for the filters

**Review by BMarink512 — APPROVED** [review]
good implementation just some minor comments

File: src/author/author.controller.ts (hunk 2 above)
Comment (AlfreMK) [reply src/author/author.controller.ts:24]: oops

File: src/utils.ts
Comment (AlfreMK) [inline src/utils.ts:1]: What do you think of this implementation? @BMarink512

Comment (AlfreMK) [reply src/utils.ts:1]: Maybe we should add tests for this in the future but right now I prefer doing other stuff tbh XD

Comment (BMarink512) [reply src/utils.ts:1]: Totally agree, it is not exactly what I had in mind, but it does the work, and the method I think looks cleaner maybe I would move the mapping outside the getAllBooks to leave it with less responsibilities

**Pull request merged**